### ISSUE TITLE
Comparison output readability

### ISSUE THREAD

**Issue by AbhiAuto** [red6/pdfcompare#59]
We have tried to compare two PDF files, and output came in the below format. It seems like we cannot read the difference from the output. Is there any other way we can make it more simpler and easy to understand?

![image](https://user-images.githubusercontent.com/26138640/64869692-784c1500-d65f-11e9-9913-f3970f790453.png)

**Comment by finsterwalder** [comment]
What do you want/expect?
Yes, it's hard to read.
But here is what pdfcompare does: It compares pdf pages pixel by pixel.
It shows pixel, that are missing in red and pixels that are there in addition in green.
When text is "moved", it quickly results in a lot of green and red, like in your example.
There is no easy way to detect a move and show that in any meaningful way, I think.

You could probably add a feature, that only shows the green **or** red.
What I usually do, when I have a big difference, like the one above is: I look at the two pdfs side by side. ;-)

**Comment by AbhiAuto** [comment]
@finsterwalder  Thanks for your suggestions :) 

I was just checking if we have a solution for getting two outputs one in green and other in red or to get a single output.

However as you mentioned, I will try to add the feature or try to find a different solution. If those didn't work I will go for the final suggestion which you have mentioned. I have to look at the two PDF files side by side :) 

One more question : Can we generate the output only with the difference into a Excel sheet or a word file ?

**Comment by HETU029** [comment]
I have not found any solution for PDF comparison result in single color. When will You add a feature, that only shows the any one color.

**Comment by finsterwalder** [comment]
BTW: It's Open Source. You can add the feature yourself and do a Pull
Request. ;-)
I don't know when or if I add this feature.

HETU029 <notifications@github.com> schrieb am Di., 17. Sep. 2019, 07:51:

**Comment by finsterwalder** [comment]
I added the ability to customize the color of diff outputs. So you could run two comparisons with different coloring. Making one color something like gray or white, might give you the effect you are looking for.
(There is not "do not change the color" option, though.)

**Comment by shahmit** [comment]
Is it possible to get the result file side by side instead of overlapping, something like what this site shows-
https://www.diffchecker.com/diff

**Comment by finsterwalder** [comment]
PdfCompare has a very simplistic GUI, that can show two files side by side.
Right now you can choose to view the expected or the actual PDF next to the compare result and switch between those.

Feel free to code something that generates two PDFs with different color markers.
something like:
new PdfComparator(file1, file2).withEnvironment(new SimpleEnvironment().setActualColor(Color.BLACK))
new PdfComparator(file1, file2).withEnvironment(new SimpleEnvironment().setExpectedColor(Color.BLACK))

And show those files.

I will most likely not work on something like this in the near future.

And I think diffchecker.com works completely different, since it seams to extract text from the PDF and does not do a pixel comparison.

**Comment by shahmit** [comment]
Thanks for the response, the result file with the differences currently gives a ‘Unified’ view which is not that readable.
Is it possible to get the result output showing the 2 files in the ‘Split’ view- which shows the 2 documents side by side?
This feature- I think would be a game changer 😊

**Comment by finsterwalder** [comment]
I'm not sure I understand what you mean.
With "side by side", I assume you mean the PdfCompare UI?
And which "2 documents" do you mean exactly?

**Comment by v3g3t4x** [comment]
check pdfcompare into my github. I added an interesting feature. For two pdf I generate one pdf where in each page you'll have on left page of pdf1 and on right a page of pdf2 with difference.

**Comment by nilimapradipm** [comment]
> check pdfcompare into my github. I added an interesting feature. For two pdf I generate one pdf where in each page you'll have on left page of pdf1 and on right a page of pdf2 with difference. check branch horizontal_compare (@finsterwalder may be an interesting feature to merge into your project).

is this implemented?

**Comment by gastendonk** [comment]
> is this implemented? side by side compariosn ?

I think it is. See class DisplayMain.

**Comment by nilimapradipm** [comment]
> > is this implemented? side by side compariosn ?
> 
> I think it is. See class DisplayMain.

Thanks for update. However i was thinking it is something which can be run programtically by passing two pdf & saving pdf for side by side comparison result

**Comment by v3g3t4x** [comment]
Sorry for delay, yes in my version on my github is implemented the side by side compare.

**Comment by nilimapradipm** [comment]
> Sorry for delay, yes in my version on my github is implemented the side by side compare.

Please can u share the class where i can call this

**Comment by v3g3t4x** [comment]
Here an example on how use it:

`			PdfComparator compare = new PdfComparator("input/input1.pdf", "input/input2.pdf");
			
			compare.headerLeft = "LEFT";
			compare.headerRight = "RIGHT";

			compare.withEnvironment(new SimpleEnvironment().setActualColor(Color.red).setParallelProcessing(true)
					.setMaxImageSize(100000).setDocumentCacheSize(0).setOverallTimeout(15)
					.setExpectedColor(Color.CYAN).setAddEqualPagesToResult(true).setDPI(DPI)
					.**setEnableHorizontalCompareOutput**(flagLeftRight).setAllowedDiffInPercent(pixel));

			CompareResult result = compare.compare();`

**Comment by nilimapradipm** [comment]
> Here an example on how use it...use **setEnableHorizontalCompareOutput**
> 
> ```
> 		PdfComparator compare = new PdfComparator("input/input1.pdf", "input/input2.pdf");
> 		boolean flagLeftRight = true;
> 		compare.headerLeft = "LEFT";
> 		compare.headerRight = "RIGHT";
> 
> 		compare.withEnvironment(new SimpleEnvironment().setActualColor(Color.red).setParallelProcessing(true)
> 				.setMaxImageSize(100000).setDocumentCacheSize(0).setOverallTimeout(15)
> 				.setExpectedColor(Color.CYAN).setAddEqualPagesToResult(true).setDPI(DPI)
> 				.setEnableHorizontalCompareOutput(flagLeftRight).setAllowedDiffInPercent(pixel));
> 
> 		CompareResult result = compare.compare();
> ```

thanks but m getting error for      below. headerLeft is not getting listed. I am on horizontal_compare branch
        compare.headerLeft = "LEFT";
        compare.headerRight = "RIGHT";

**Comment by v3g3t4x** [comment]
Yes sorry, my error.
You can now find the fix into horizontal_compare branch. Tell me if other error occurs.

**Comment by v3g3t4x** [comment]
Pushed new code. Thanks
Pull

**Comment by ayush758** [comment]
@v3g3t4x If I have a multipage PDF it compares the page and overwrites the output file, Is there a way to output multiple files for each page it compares

**Comment by v3g3t4x** [comment]
Hi,
It create one PDF with multipage. Check well please and tell me if the issue is real.
But if I don't remember wrong it create one PDF with N pages.

**Comment by LordOfStrings** [comment]
@v3g3t4x it creates a single png file called comp.png rather than a pdf file and that png file is replaced over and over again fo reach comparison it does, do I need to change the format of the png file to PDF? I am using the horizontal_compare branch.

**Comment by v3g3t4x** [comment]
That png it's only a tmp file, it produce a PDF too with all pages

**Comment by LordOfStrings** [comment]
Can you list me to the class where it creates the pdf file?  because in the output I don't see any pdf files.

**Comment by LordOfStrings** [comment]
@v3g3t4x can you list me to the piece of code where it does the same, the only output I am getting is in the comp.png ( do I have to look in any specific folder)

**Comment by finsterwalder** [comment]
It's kinda hard to follow what exactly you are doing.
PdfCompare does not write a result Pdf, unless you tell it to do so.
For that you need to call CompareResult.writeTo(). This will store a multi page Pdf with one page per page from the original documents, marking all the differences.
There is no feature to write a single compare file for each page.

**Comment by v3g3t4x** [comment]
Exactly, thanks @finsterwalder

**Comment by heenavasdani1** [comment]
@finsterwalder , Can you please look into PR which @v3g3t4x created to merge ? He implemented a feature for comparasion where he generated one pdf where in each page you'll have on left page of pdf1 and on right a page of pdf2 with difference. 
https://github.com/red6/pdfcompare/issues/59#issuecomment-660713200

**Comment by heenavasdani** [comment]
https://github.com/red6/pdfcompare/issues/59#issuecomment-2608976696

 @rweisleder , @pSub , @SoltauFintel
